src/test/librados/tier_cxx.cc

```diff
@@ -155,6 +155,47 @@ string get_fp_oid(string oid, std::string fp_algo = NULL)
   return string();
 }
 
+bool is_intended_refcount_state(librados::IoCtx& src_ioctx,
+				std::string src_oid,
+				librados::IoCtx& dst_ioctx,
+				std::string dst_oid,
+				uint64_t expected_refcount)
+{
+  int src_refcount = 0, dst_refcount = 0;
+  {
+    bufferlist t;
+    int r = dst_ioctx.getxattr(dst_oid, CHUNK_REFCOUNT_ATTR, t);
+    if (r == -ENOENT) {
+      dst_refcount = 0;
+    } else {
+      chunk_refs_t refs;
+      try {
+	auto iter = t.cbegin();
+	decode(refs, iter);
+      } catch (buffer::error& err) {
+	ceph_assert(0);
+      }
+      dst_refcount = refs.count();
+    }
+  }
+  if (dst_refcount == expected_refcount) {
+    return true;
+  }
+  {
+    bufferlist in, out;
+    encode(dst_oid, in);
+    src_refcount = src_ioctx.exec(src_oid, "cas", "references_chunk", in, out);
+    if (src_refcount == -ENOENT || src_refcount == -ENOLINK) {
+      src_refcount = 0;
+    }
+    ceph_assert(src_refcount >= 0);
+  }
+  if (src_refcount > dst_refcount) {
```

Comment:
Isn't this backwards?  The design is that the refcount recorded on the target is always >= than the real number of references, right?  We always increment the refcount on the target before updating the source metadata and update the source metadata before decrementing the target refcount, right?

Reply:
src_refcount should always match expected_refcount, right?  I think you should always read dst_refcount and src_refcount and assert:
1. src_refcount == expected_refcount
2. dst_refcount >= src_refcount

Reply:
src means here is the manifest object on the base tier, and dst means that the the chunked object on the low tier.
So, isn't dst_refcount == expected_refcount correct?

Reply:
I don't think so, we specifically allow dst_refcount to be larger.

Reply:
Yeah, dst_refcount >= src_refcount is correct.
but expected_refcount here is to check whether dst_refcount is the refcount as our expectation or not.

Reply:
Oh, sorry, I misunderstood. assert:

    src_refcount == expected_refcount
    dst_refcount >= src_refcount

are correct. Can you take a look a commit I added?